backend/test_observer/controllers/test_results/test_results.py

```diff
@@ -0,0 +1,307 @@
+
+    if until:
+        filters.append(TestExecution.created_at <= until)
+
+    # Build the main query with window function for total count in a single query
+    query = (
+        select(TestResult, func.count().over().label("total_count"))
+        .join(TestResult.test_execution)
+        .join(TestExecution.environment)
+        .join(TestExecution.artefact_build)
+        .join(ArtefactBuild.artefact)
+        .join(TestResult.test_case)
+        .options(
+            joinedload(TestResult.test_execution).joinedload(TestExecution.environment),
+            joinedload(TestResult.test_execution)
+            .joinedload(TestExecution.artefact_build)
+            .joinedload(ArtefactBuild.artefact),
+            joinedload(TestResult.test_case),
+        )
+        .order_by(desc(TestExecution.created_at), desc(TestResult.id))
+        .offset(offset)
+        .limit(limit)
+    )
+
+    # Apply all filters
+    if filters:
+        query = query.where(and_(*filters))
+
+    result = db.execute(query).all()
+
+    if result:
+        test_results = [row[0] for row in result]
+        total_count = result[0][1]
+    else:
+        test_results = []
+        total_count = 0
+
+    return TestResultSearchResponse(count=total_count, test_results=test_results)
+
+
+@router.get("/families", response_model=list[str])
+def get_families(db: Session = Depends(get_db)) -> list[str]:
+    """
+    Returns all families that have test results.
+    """
+    families = (
+        db.execute(
+            select(distinct(Artefact.family))
+            .join(ArtefactBuild)
+            .join(TestExecution)
+            .join(TestResult)
+            .order_by(Artefact.family)
+        )
+        .scalars()
+        .all()
+    )
+
+    return [family.value for family in families]
+
+
+@router.get("/environments", response_model=list[str])
+def get_environments(
+    family: str | None = Query(None, description="Filter environments by family"),
+    test_case: str | None = Query(None, description="Filter environments by test case"),
+    db: Session = Depends(get_db),
+) -> list[str]:
+    """
+    Returns list of distinct environments, optionally filtered by other criteria.
+    """
+
+    query = (
+        select(distinct(Environment.name))
+        .join(TestExecution)
+        .join(TestResult)
+        .join(TestExecution.artefact_build)
+        .join(ArtefactBuild.artefact)
+        .join(TestResult.test_case)
+    )
+
+    # Apply cascading filters
+    if family:
+        try:
+            family_enum = FamilyName(family.lower())
+            query = query.where(Artefact.family == family_enum)
+        except ValueError:
+            return []  # Invalid family
+
+    if test_case:
+        query = query.where(TestCase.name.ilike(f"%{test_case}%"))
+
+    query = query.order_by(Environment.name)
+
+    environments = db.execute(query).scalars().all()
+    return list(environments)
+
+
+@router.get("/test-cases", response_model=list[str])
```

Comment:
Like above, I think this should be moved to `GET /v1/test-cases` and the ability to filter removed.